File: openbank-admin-ui/src/lib/validation/iban.ts
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) OpenBank contributors. Licensed under the Mozilla Public License 2.0.
// See LICENSE in the repository root or https://www.mozilla.org/MPL/2.0/ for details.

// Client-side IBAN validation. Why this exists: the account-service rejects any
// non-IBAN string at the domain boundary with a raw `Invalid IBAN: <value>`
// (openbank-libs Iban.kt). If the operator types `*`, `CZ` or any free-text into
// an IBAN-bound field, that raw exception used to surface verbatim in the UI. We
// validate the shape *before* calling the BFF so a malformed value never reaches
// the IBAN endpoint — the user gets an inline hint instead of a leaked backend
// error. ISO 13616: 2 country letters + 2 check digits + up to 30 BBAN chars,
// verified with the mod-97 checksum.

/** Strip spaces and upper-case. IBANs are printed in groups of 4 but stored flat. */
export function normalizeIban(raw: string): string {
  return raw.replace(/\s+/g, '').toUpperCase()
}

/** Cheap structural check: country code + check digits + alphanumeric BBAN. */
export function hasIbanShape(raw: string): boolean {
  return /^[A-Z]{2}\d{2}[A-Z0-9]{10,30}$/.test(normalizeIban(raw))
}

/** RFC 4122 canonical UUID (account-service identifies a party by UUID). */
export function looksLikeUuid(raw: string): boolean {
  return /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i.test(raw.trim())
}

/** Full ISO 13616 mod-97 validation (shape + checksum). */
export function isValidIban(raw: string): boolean {
  const v = normalizeIban(raw)
  if (!hasIbanShape(v)) return false
  // Move the first four chars to the end, then replace letters with 2-digit codes.
  const rearranged = v.slice(4) + v.slice(0, 4)
  const numeric = rearranged.replace(/[A-Z]/g, ch => String(ch.charCodeAt(0) - 55))
  // mod-97 over a long numeric string, processed in chunks to avoid BigInt overflow.
  let remainder = 0
  for (let i = 0; i < numeric.length; i += 7) {
    remainder = Number(`${remainder}${numeric.slice(i, i + 7)}`) % 97
  }
  return remainder === 1
}

File: openbank-libs/src/main/kotlin/com/openbank/libs/domain/account/Iban.kt
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) OpenBank contributors. Licensed under the Mozilla Public License 2.0.
// See LICENSE in the repository root or https://www.mozilla.org/MPL/2.0/ for details.

package com.openbank.libs.domain.account

data class Iban(val value: String) {
    init {
        require(isValid(value)) { "Invalid IBAN: $value" }
    }

    val countryCode: String get() = value.substring(0, 2)
    val checkDigits: String get() = value.substring(2, 4)
    val bban: String get() = value.substring(4)

    fun formatted(): String = value.chunked(4).joinToString(" ")

    override fun toString(): String = value

    companion object {
        fun of(raw: String): Iban = Iban(raw.replace(" ", "").uppercase())

        fun isValid(iban: String): Boolean {
            val cleaned = iban.replace(" ", "").uppercase()
            if (cleaned.length < 15 || cleaned.length > 34) return false
            if (!cleaned.substring(0, 2).all { it.isLetter() }) return false
            if (!cleaned.substring(2, 4).all { it.isDigit() }) return false
            val rearranged = cleaned.substring(4) + cleaned.substring(0, 4)
            val numeric = rearranged.map { c ->
                if (c.isDigit()) c.toString() else (c.code - 'A'.code + 10).toString()
            }.joinToString("")
            return numeric.toBigInteger().mod(97.toBigInteger()).toInt() == 1
        }
    }
}
